### PR TITLE
better cleaning of urls and enable local configs

### PR DESCRIPTION
**Pull request by queirozfcom**
1) make the generated urls for anchors be more like slugs (e.g. "/" also becomes a dash)
2) I frequently want some posts to have more extensive TOCs than others. I've changed the rendering and the filtering so that you can also define a front matter as follows, and these local configurations are only valid for the current post:

```
---
layout: page
header: no
title: "Post title here"
toc: 
    max_level: 3
---
post contents here.
for this post only, the toc configs will have "max_level" => 3, so that TOCs up to header-3 will be printed.
```

### CODE REVIEW

**Comment by queirozfcom** [comment]
See an example on my site:
- post with max_level = 3 (local config overrides global config): http://queirozf.com/entries/matplotlib-pylab-pyplot-etc-what-s-the-different-between-these
- post with max_level = 2 (global config setting): http://queirozf.com/entries/podcast-episode-overview-what-machine-learning-engineers-need-to-know

File: lib/table_of_contents/parser.rb [hunk 1]
```diff
@@ -44,8 +44,11 @@ def parse_content
         @doc.css(toc_headings).each do |node|
           text = node.text
           id = text.downcase
-          id.gsub!(PUNCTUATION_REGEXP, '') # remove punctuation
+          id.gsub!(PUNCTUATION_REGEXP, ' ') # replace punctuation with spaces
+          id.gsub!(/\s+/,' ') # in case there were two or more special characters
           id.gsub!(' ', '-') # replace spaces with dash
+          id.gsub!(/-+/, '-') # in case there were two or more dashes
+          id.chomp!('-') # not at the end
```
Comment (toshimaru) [inline lib/table_of_contents/parser.rb]: Why did you change this id generation logic??

I tried the below toc generation and it does work. 

## heading

```

## a b  c   d

## foo-bar

## foo bar

## a--b---c
```

result:

```
<ul class="section-nav">
<li class="toc-entry toc-h2"><a href="#a-b--c---d">a b  c   d</a></li>
<li class="toc-entry toc-h2"><a href="#hoge-foo">hoge-foo</a></li>
<li class="toc-entry toc-h2"><a href="#hoge-foo-1">hoge foo</a></li>
<li class="toc-entry toc-h2"><a href="#abc">a–b—c</a></li>
<li class="toc-entry toc-h2"><a href="#a-b-c">a b c</a></li>
</ul>
``` 

What toc case do you want to care?

Comment (queirozfcom) [reply lib/table_of_contents/parser.rb]: I think it breaks in cases when the title has, for example, a slash (`/`) or a question mark (`?`)

Comment (toshimaru) [reply lib/table_of_contents/parser.rb]: I think it works although the title has slash or question mark.

```
## abc

## a/b/c

## abc?
```

```html
<li class="toc-entry toc-h2"><a href="#abc">abc</a></li>
<li class="toc-entry toc-h2"><a href="#abc-1">a/b/c</a></li>
<li class="toc-entry toc-h2"><a href="#abc-2">abc?</a></li>
```

This works correctly.

**Comment by toshimaru** [comment]
If you want to ignore the specific headings, you can use the feature: [skip-toc](https://github.com/toshimaru/jekyll-toc#skip-toc).

So, let me close this PR. If any further feature request, please let me know. 😃
